src/widget/wbpmeditor.cpp

```diff
@@ -0,0 +1,254 @@
+}
+
+WBpmEditor::WBpmEditor(const QString& group, QWidget* pParent)
+        : WWidget(pParent),
+          m_group(group),
+          m_tempoTapCO(m_group, QStringLiteral("tempo_tap")),
+          m_bpmTapCO(m_group, QStringLiteral("bpm_tap")),
+          // TODO ControlProxy, quit editor when track is unloaded?
+          m_trackLoadedCO(m_group, QStringLiteral("track_loaded")),
+          m_bpmCO(m_group, QStringLiteral("bpm")),
+          m_fileBpmCO(m_group, QStringLiteral("file_bpm")),
+          m_rateRatioCO(m_group, QStringLiteral("rate_ratio")),
+          m_pClickOverlay(nullptr),
+          m_pSelectWidget(nullptr),
+          m_pTapSelectButton(nullptr),
+          m_pEditSelectButton(nullptr),
+          m_pTapButton(nullptr),
+          m_pEditBox(nullptr) {
+    setContentsMargins(0, 0, 0, 0);
+    // Setup the layout with the mode selectors (Tap|Edit), tap button and
+    // tempo editor. Only one of these widgets can be visible.
+    m_pClickOverlay = std::make_unique<QPushButton>(this);
+    m_pClickOverlay->setObjectName(QStringLiteral("BpmClickButton"));
+    m_pClickOverlay->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);
+    m_pClickOverlay->setFocusPolicy(Qt::NoFocus);
+    m_pClickOverlay->installEventFilter(this);
+    connect(m_pClickOverlay.get(),
+            &QPushButton::clicked,
+            this,
+            [this]() { switchMode(Mode::Select); });
+
+    m_pTapSelectButton = std::make_unique<QPushButton>(this);
+    m_pTapSelectButton->setObjectName(QStringLiteral("BpmTapSelectButton"));
+    m_pTapSelectButton->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);
+    m_pTapSelectButton->setFocusPolicy(Qt::NoFocus);
+    m_pTapSelectButton->installEventFilter(this);
+    connect(m_pTapSelectButton.get(),
+            &QPushButton::clicked,
+            this,
+            [this]() { switchMode(Mode::Tap); });
+
+    m_pEditSelectButton = std::make_unique<QPushButton>(this);
+    m_pEditSelectButton->setObjectName(QStringLiteral("BpmEditSelectButton"));
+    m_pEditSelectButton->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);
+    m_pEditSelectButton->setFocusPolicy(Qt::NoFocus);
+    m_pEditSelectButton->installEventFilter(this);
+    connect(m_pEditSelectButton.get(),
+            &QPushButton::clicked,
+            this,
+            [this]() { switchMode(Mode::Edit); });
+
+    m_pSelectWidget = std::make_unique<QWidget>(this);
+    m_pSelectWidget->installEventFilter(this);
+    auto pSelectLayout = new QHBoxLayout();
+    pSelectLayout->setAlignment(Qt::AlignHCenter | Qt::AlignVCenter);
+    pSelectLayout->setContentsMargins(0, 0, 0, 0);
+    pSelectLayout->setSpacing(0);
+    pSelectLayout->setSizeConstraint(QLayout::SetNoConstraint);
+    pSelectLayout->addWidget(m_pTapSelectButton.get());
+    pSelectLayout->addWidget(m_pEditSelectButton.get());
+    pSelectLayout->update();
+    pSelectLayout->activate();
+    m_pSelectWidget->setLayout(pSelectLayout);
```

Comment:
when paired with `make_unique` above
```suggestion
    m_pSelectWidget->setLayout(pSelectLayout.release());
```